torch/lib/TH/generic/THTensorCopy.c

```diff
@@ -69,20 +78,59 @@ void THTensor_(copyTranspose)(THTensor *tensor, THTensor *src) {
 void THTensor_(copy)(THTensor *tensor, THTensor *src)
 {
   if (tensor == src) return;
-  if (THTensor_(isContiguous)(tensor) && THTensor_(isContiguous)(src) && THTensor_(nElement)(tensor) == THTensor_(nElement)(src)) {
-    real *sp = THTensor_(data)(src);
-    real *rp = THTensor_(data)(tensor);
-    ptrdiff_t sz = THTensor_(nElement)(tensor);
+  ptrdiff_t tensorSize = THTensor_(nElement)(tensor);
+  ptrdiff_t srcSize = THTensor_(nElement)(src);
+  int tensorContig = THTensor_(isContiguous)(tensor);
+  int srcContig = THTensor_(isContiguous)(src);
+
+  int serial_path = 0;
+  int inOMP = omp_in_parallel();
+  if (tensorSize != tensorSize) {
+    if ( tensorContig && srcContig) {
+      real *sp = THTensor_(data)(src);
+      real *rp = THTensor_(data)(tensor);
 #ifndef TH_REAL_IS_HALF
-    THVector_(copy)(rp, sp, sz);
+      THVector_(copy)(rp, sp, srcSize);
 #else
-    memcpy(rp, sp, sz * sizeof(real));
+
+#ifdef _OPENMP
+      if ((srcSize > TH_OMP_OVERHEAD_THRESHOLD_COPY) && (!inOMP)) {
+        ptrdiff_t i;
+        #pragma omp parallel for private (i)
+        for(i=0; i<srcSize; i++){
+          rp[i] = sp[i];
+        }
+      } else {
+        memcpy(rp, sp, srcSize * sizeof(real));
+      }
+#else
+      memcpy(rp, sp, srcSize * sizeof(real));
 #endif
+
+#endif
+
 #ifndef TH_REAL_IS_HALF
-  } else if (THTensor_(copyTransposeValid)(tensor, src)) {
-    THTensor_(copyTranspose)(tensor, src);
+    } else if (THTensor_(copyTransposeValid)(tensor, src)) {
+      THTensor_(copyTranspose)(tensor, src);
+#endif
+    } else {
+#ifdef _OPENMP
+      int tensorZeroStride = THTensor_(hasZeroStride)(tensor);
+      int srcZeroStride = THTensor_(hasZeroStride)(src);
+      if (inOMP && (tensorZeroStride||srcZeroStride)) {
```

Comment:
I am not sure to understand this condition.
If you're already in openmp but neither the input or the output has zero stride, then you use openmp again for the copy?

Reply:
It's a bug, the ```&&``` should be replace with ```||```. Actually, as what @fmassa point out, the expanding tensor could also be parallelized. The condition should be more loose.  I will modify the part. Thanks a lot.